### ISSUE TITLE
"Too small initial heap" maybe introduced by #365

### ISSUE THREAD

**Issue by Nerus92** [itzg/docker-minecraft-server#366]
Hello!

It would appear that #365 brought some issues to my setup and I would like to help determine if that is a problem that only I am experiencing or a more general one.

I have been running a Forge server for several months now, on a GCP compute instance n1-highmem-2 (2 vCPUs, 13 GB memory). 
The environment variables are as followed:
VERSION	1.12.2
TYPE	FTB
FTB_SERVER_MOD	https://minecraft.curseforge.com/projects/enigmatica2/files/2672373
EULA	TRUE
VIEW_DISTANCE	12
SEED	-1183769889
LEVEL_TYPE	BIOMESOP
ALLOW_FLIGHT	TRUE
INIT_MEMORY	8G
MAX_MEMORY	10G
The container is linked to a SSD volume for persistence.

I have been starting and stopping the instance every time I started and stopped playing, and its been perfect up until today. **I have been able to use the instance without any issue yesterday night.**

As I started the instance today, I saw the attached stackdriver logs. 
**tldr**: a reboot loop of the container around a "Too small initial heap" message.

Here is an extract of it for ease:

```
A 2019-08-01T04:10:13.942941153Z Starting server
A 2019-08-01T04:10:13.953817147Z Error occurred during initialization of VM
A 2019-08-01T04:10:13.954827375Z Too small initial heap
A 2019-08-01T04:10:13.957134354Z Server will restart in ~10 seconds. No input needed...
A 2019-08-01T04:10:13.960542957Z Ping to Google DNS successfull
A 2019-08-01T04:10:13.962682359Z Ping to L4 successfull
A 2019-08-01T04:10:13.964071998Z Rebooting now!
```

Looking for the source of the issue and knowing that I didn't change my instance's configuration since yesterday, I realized that there had been a new docker image published earlier today from #365.
To confirm that the issue came from the PR and not from my own environment/configuration, I restarted the instance using itzg/minecraft-server:multiarch (instead of latest) which is from the commit just preceding the PR (5 days ago). The instance started normally, the server went up, and I was able to join the server as per usual. **Reverting back to itzg/minecraft-server:latest produces the result from the logs again.**

As additional information, I tried using the MEMORY environment variable to change from the default to something bigger (2G and 4G) but to no avail, the same error causes the reboot loop.

Please let me know if any additional information would help.

[too_small_initial_heap-minecraft-logs.txt](https://github.com/itzg/docker-minecraft-server/files/3454871/too_small_initial_heap-minecraft-logs.txt)

**Comment by iamtpage** [comment]
Try:

```
      INIT_MEMORY: "8G"
      MAX_MEMORY: "10G"
```

And drop the `MEMORY` variable

**Comment by iamtpage** [comment]
Now that I've had my coffee, what does `settings.cfg` look like? (`/data/FeedTheBeast/settings.cfg`)

**Comment by itzg** [comment]
Strange, but yeah, definitely seems to be something about that particular change...which should have been for the better :)

I'm also curious about `settings.cfg` to see what got produced there.

As I mentioned in #357 , there might be a more general problem with trying to leverage the non-standard-enforced startup scripts of FTB and CurseForge modpacks. I'm thinking about leveraging this existing code

https://github.com/itzg/docker-minecraft-server/blob/f8b97bf90984b3357beab4f3ade5211793b9f43e/start-deployFTB#L77-L93

to have it fabricate a startup script regardless of whether one was provided by the modpack.

**Comment by jpchatainsbra** [comment]
Thank you for looking into it!

Please find attached the full settings.cfg for the healthy and unhealthy version (obtained by restarting the instance and pointing, respectively, to itzg/minecraft-server:multiarch and itzg/minecraft-server).

And here is the diff:
```
Nerus92$ diff unhealthy_settings-cfg.txt healthy_settings-cfg.txt
4c4
< MAX_RAM=10G
---
> MAX_RAM=10G;
```

Let me know if any other log / trace / ... would help.

Relative to the note in #357, I can give an additional data point which is that the FTB_SERVER_MOD environment variable stayed the same since the creation of the instance a few months back. So hopefully no change on any provided startup script. 
I know nothing to how modpacks work, do they do their own parsing of the config files? If this one was to actually rely on a specific format of settings.cfg and more particularly on a semi-colon being present at the end of each line (which seems to be the case for all the other lines of the file) that might explain the situation.

[healthy_settings-cfg.txt](https://github.com/itzg/docker-minecraft-server/files/3458407/healthy_settings-cfg.txt)
[unhealthy_settings-cfg.txt](https://github.com/itzg/docker-minecraft-server/files/3458408/unhealthy_settings-cfg.txt)

**Comment by iamtpage** [comment]
You're `settings.cfg` has a lot of extraneous `;` at the end of the lines. Take a look at the sample one from the source repository:
https://github.com/AllTheMods/Server-Scripts/blob/master/settings.cfg

As you can see, no semicolons. Try removing them from the `settings.cfg` and see what happens

**Comment by itzg** [comment]
@iamtpage this is going to sound bizarre, but the trailing semicolons are actually expected/required. The modpack's startup script has this very particular way of reading `settings.cfg`:

```
read_config(){
        while read -r line || [[ -n "$line" ]] ; do
                if echo $line | grep -F = &>/dev/null; then
                        if [[ ${str:0:1} != "#" ]] ; then
                        name=$(echo "$line" | cut -d '=' -f 1)
                        val=$(echo "${line}" | cut -d '=' -f 2-)
                        eval "export ${name}='${val%?}'"
                fi
                fi
        done < settings.cfg

}
```

where the `${val%?}` is actually trimming the last character of the value...the semicolon. Here's a small snippet where I experimented with the logic there:

```
bash-4.4# val=$(echo "MAX_RAM=1G" | cut -d '=' -f 2-)
bash-4.4# echo ${val%?}
1
```

I'm afraid #365 needs to be reverted even though in itself made total sense.

What modpack were you using that led to #364 / #365?

**Comment by itzg** [comment]
Fixing with revert in e61942f33e504495271eb9152edf4e35983ee152

**Comment by itzg** [comment]
Image with `sha256:510b34f61b3b4d5eb5e46b802f2053e676028fd9c142917f8524facb69924e4c` is now available on Docker Hub.
